### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by mikluke**

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	}
+}
+
+func WithErrHandler(f func(height uint64, err error)) FetchBlockOption {
+	return func(opts *FetchBlockOptions) {
+		opts.errHandler = f
+	}
+}
+
+func WithContext(ctx context.Context) FetchBlockOption {
```
Comment (megaherz) [inline options.go]: How about making context as a mandatory parameter? It is more common.

File: options.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package ariadne
+
+import (
+	"context"
+	"time"
+)
+
```
Comment (megaherz) [inline options.go:6]: Just an idea maybe use an alias as `type Height = uint64`

File: README.md [hunk 3]
```diff
@@ -1 +1,30 @@
...
+## Usage
+
+Short example:
+
+```go
+f, err := ariadne.New(nodeAddr, cdc, time.Minute)
+if err != nil { panic(err) }
+
+for b := range f.FetchBlocks(b.Height) {
+    spew.Dump(b)
```
Comment (megaherz) [inline README.md]: spew is a 3rd party lib, maybe something as `fmt.Printf("%+v\n", b)`

File: fetcher.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	for _, tx := range b.Txs {
+		for _, msg := range tx.GetMsgs() {
+			msgs = append(msgs, msg)
+		}
+	}
+
+	return msgs
+}
+
+func FilterMessages(msgs []sdk.Msg, types ...reflect.Type) []sdk.Msg {
```
Comment (megaherz) [inline fetcher.go]: How about FilterMessages as an option?

File: ariadne.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+type Block struct {
+	Height uint64
+	Txs    []sdk.Tx
+}
+
+//go:generate mockgen -destination=./mock/ariadne_mock.go -package=mock -source=ariadne.go
+
+// Fetcher interface for fetching.
+type Fetcher interface {
+	FetchBlocks(from uint64, opts ...FetchBlocksOption) <-chan Block
```
Comment (megaherz) [inline ariadne.go]: Maybe return *Block for both FetchBlocks and FetchBlock?

File: options.go (hunk 1 above)
Comment (mikluke) [reply options.go]: Good point

File: options.go (hunk 2 above)
Comment (mikluke) [reply options.go:6]: We force users to do redundant cast, I won't do it

File: README.md (hunk 3 above)
Comment (mikluke) [reply README.md]: okay

File: fetcher.go (hunk 4 above)
Comment (mikluke) [reply fetcher.go]: It's not obvious how it should work. We should cut messages from transactions? or how? I tried to wrap FetchBlocks with FetchMessages but this way doesn't allow to use options. Or we should define types parameter as obligatory

File: README.md [hunk 6]
```diff
@@ -1 +1,32 @@
...
+}
+````
+
+You have to look at detailed example [there](example/example.go)
```
Comment (megaherz) [inline README.md]: there => here :-)

File: ariadne.go (hunk 5 above)
Comment (mikluke) [reply ariadne.go]: Done for FetchBlock

**Pull request merged**